### ISSUE TITLE
Method error with generic floats

### ISSUE THREAD

**Issue by urchgene** [timholy/PositiveFactorizations.jl#19]
Hi,

First, I really like PositiveFactorizations. I cannot do anything without cholesky factorization and so with non PD matrices which is very common in genetics and breeding, this package comes very handy.

However, I get method error when using it with PDMats for MvNormal.

```
Closest candidates are:
  floattype(!Matched::Type{T<:AbstractFloat}) where T<:AbstractFloat at /Users/ugokek/.julia/packages/PositiveFactorizations/VybOm/src/cholesky.jl:264
  floattype(!Matched::Type{T<:Integer}) where T<:Integer at /Users/ugokek/.julia/packages/PositiveFactorizations/VybOm/src/cholesky.jl:265
```

I have cutout some of the error msg cos we figured it was coming from PositiveFactorizations.

The issue is here if you wanna see the entire thing:
https://github.com/tpapp/DynamicHMCExamples.jl/issues/5

Thanks.

Uche.

**Comment by timholy** [comment]
Without a *minimal* example that I can cut and paste, there's nothing I can do. But if you have an unusual number type you could always try defining a specialization of `PositiveFactorizations.floattype`:

```julia
PositiveFactorizations.floattype(::Type{MyWeirdNumberType}) = Float64
```

That should work if `convert(Float64, x::MyWeirdNumberType)` works.

**Comment by urchgene** [comment]
Thanks a lot @timholy  for prompt reply.

I am still new to Julia so I am not sure how to do some things.

This is the basic framework:

```
G = rand(4,200); G = G*G'/200;
Vu = 0.25*G
Ve = 0.45*G

L = rand(20,1000); L = L*L'/1000; L = cholesky(L).L;

Gcov = kron(Vu, L);
Rcov = kron(Ve, zeros(20,20) + 1.0*I);

X = ones(20,1);
β = rand(4,1);

mu = vec(X*β')    
Rsig = Gcov + Rcov + 0.01*I;
    
xxx = cholesky(Positive, Rsig)
Rsig = PDMat(Rsig, xxx)
    
loglik = MvNormal(mu, Rsig)
```
This works but when we fold it into forwarddiff with @tpapp's DynamicHMC's program... it fails.

I like the solution you proposed but how do I fit the convert into here cos I'm not reading my {MyWeirdNumberType} from input.  It's here:
```
xxx = cholesky(Positive, Rsig)
```

Thanks again.

Uche.

**Comment by tpapp** [comment]
@urchgene: here is how you can make an MWE out of it using `ForwardDiff`:
```julia
using LinearAlgebra, PositiveFactorizations
import Random, ForwardDiff

Random.seed!(1)

G = rand(4,200)
G = G*G'/200;
L = rand(20,1000)
L = L*L'/1000

function cholesky_part(G, L)
    Vu = 0.25*G
    Ve = 0.45*G
    L = cholesky(L).L;
    Gcov = kron(Vu, L);
    Rcov = kron(Ve, zeros(20,20) + 1.0*I);
    Rsig = Gcov + Rcov + 0.01*I;
    cholesky(Positive, Rsig).L[1] # just for the MWE
end

ForwardDiff.jacobian(x -> cholesky_part(reshape(x, size(G)...), L), vec(G))
```

**Comment by urchgene** [comment]
Thanks @tpapp for the reply but I maybe think I miscommunicated.

I want to be able to use PositiveFactorizations with DynamicHMC.

The crux is here:

```
function (problem::mvLMM)(θ)
    
    @unpack Y, X, Z, L = problem # extract the data
    @unpack β, Omegau, Omegae, sigmau, sigmae = θ     # extract the parameters
    N = size(L, 1); d = size(Y, 1);

    Vu = getcov(sigmau, Omegau)
    Ve = getcov(sigmae, Omegae)

    Gcov = kron(Vu, L);
    Rcov = kron(Ve, zeros(N,N) + 1.0*I);

    mu = vec(X*β')    
    Rsig = Gcov + Rcov + 0.01*I;
    xxx = cholesky(Positive, Rsig).L
    loglik = AltMvNormal(Val(:L), mu, xxx)
    vecY = vec(Y); 
    loglik = logpdf(loglik, vecY) ## loglik fn must be 2-dim
    logpri = sum(logpdf.(Cauchy(2.5), sigmau)) + sum(logpdf.(Cauchy(2.5), sigmae)) + logpdf(LKJL(d), Omegau) + logpdf(LKJL(d), 
Omegae)
    loglik + logpri
end
```

This spills out error that relates to the generic float problem with PositiveFactorizations.

@timholy suggested something but I do not know yet how to use the Type convert method.

Thanks again for your responses, very useful.

Uche.

**Comment by timholy** [comment]
This may have been fixed by #27

**Comment by timholy** [comment]
I'm going to guess yes, re-open with a MWE if not.
